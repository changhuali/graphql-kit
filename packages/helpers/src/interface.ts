import type { ConstDirectiveNode, OperationTypeNode } from 'graphql'

export type Directive = ConstDirectiveNode

export type Directives = ReadonlyArray<Directive>

export type EnumTypeDef = {
  name: string
  description: string
  directives: Directives
  value: unknown
}[]

export type FieldTypeDef = {
  description: string
  directives: Directives
  type: string
  typeDef: TypeDef
}

export type UnionTypeDef = Record<string, FieldTypeDef>

export type TypeDef = EnumTypeDef | UnionTypeDef | undefined

export interface OperationArgument extends FieldTypeDef {
  name: string
}

export interface OperationReturn extends FieldTypeDef {
  name: string
}

export interface Operation {
  name: string
  description: string
  directives: Directives
  arguments: OperationArgument[]
  argumentsExample: Record<string, unknown>
  return: OperationReturn
  returnExample: Record<string, unknown>
}

export interface TypedOperation extends Operation {
  type: OperationTypeNode
}

export type ScalarMap = Record<string, unknown>
