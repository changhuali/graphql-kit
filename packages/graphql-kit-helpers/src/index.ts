export * from './operation'
export * from './gql'
export * from './interface'
