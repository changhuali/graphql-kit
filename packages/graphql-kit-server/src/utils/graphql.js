import { getNamedType, isEnumType, isScalarType, isUnionType } from "graphql"

/**
 * extract field info from field map
 * @param {import("graphql").GraphQLFieldMap<any, any> | import("graphql").GraphQLInputFieldMap} fieldMap - input definition node
 * @returns {any}
 */
export function extractFieldInfoFromFieldMap(fieldMap) {
  /** @type {Record<string, any>} */
  const result = {}
  Object.keys(fieldMap).forEach((key) => {
    const field = fieldMap[key]
    const namedType = getNamedType(field.type)
    result[key] = {
      description: field.description || '',
      type: field.type.toString(),
      directives: field.astNode?.directives,
      ...getTypeDefFromNamedType(namedType)
    }
  })
  return result
}

/**
 * extract field info from named type
 * @param {import("graphql").GraphQLNamedType} namedType - named type
 */
export function getTypeDefFromNamedType(namedType) {
  /** @type {Record<string, any>} */
  const result = {}
  if (isEnumType(namedType)) {
    result.typeDef = namedType.getValues().map(item => ({
      name: item.name,
      description: item.description || '',
      directives: item.astNode?.directives || [],
      value: item.value,
    }))
  } else if (isUnionType(namedType)) {
    result.typeDef = namedType.getTypes().map(item => {
      return extractFieldInfoFromFieldMap(item.getFields())
    }).reduce((result, item) => {
      return {...result, ...item}
    }, {})
  } else if (isScalarType(namedType)) {
    // Do nothing
  } else {
    result.typeDef = extractFieldInfoFromFieldMap(namedType.getFields())
  }
  return result
}

/**
 * extract operation info from graphql field
 * @param {import('graphql').GraphQLField<any, any>} graphQLField 
 * @param {import('graphql').GraphQLSchema} rawSchema 
 * @returns {object}
 * TODO:avoid circular ref
 */
 export function extractOperationInfoFromGraphQLField(graphQLField, rawSchema) {
  const typeMap = rawSchema.getTypeMap()
  const returnTypeValue = typeMap[getNamedType(graphQLField.type).name]
  return {
    name: graphQLField?.name,
    description: graphQLField?.description || '',
    directives: graphQLField.astNode?.directives || [],
    arguments: graphQLField?.args.map(item => {
      const namedType = getNamedType(item.type)
      const namedTypeValue = typeMap[namedType.name]
      return {
        name: item.name,
        description: item.description || '',
        directives: item.astNode?.directives || [],
        type: item.type.toString(),
        ...getTypeDefFromNamedType(namedTypeValue),
      }
    }),
    return: {
      name: graphQLField.name,
      directives: returnTypeValue.astNode?.directives,
      description: returnTypeValue.astNode?.description?.value || '',
      type: graphQLField?.type,
      ...getTypeDefFromNamedType(returnTypeValue),
    },
  }
}
