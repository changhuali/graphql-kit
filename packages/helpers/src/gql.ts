import type { ArgTypeDef, ObjectFieldTypeDef, ObjectTypeDef } from '.'
import type { TypedOperation } from './interface'

/**
 * generate a space string
 * @param count - the str length
 */
export const genSpace = (count: number) => {
  return ' '.repeat(count)
}

/**
 * generate a typeDef gql
 * @param typeDef - the type of operation
 * @param indent - the indent of gql
 */
const _genObjectFieldTypeDefStr = (
  objectField: ObjectFieldTypeDef,
  lineIndent: number,
  indent: number,
  allArgs: ArgTypeDef[],
) => {
  const { name, args, output } = objectField
  let str = `${genSpace(lineIndent)}${name}`
  if (args.length) {
    str += `(${args
      .map(item => {
        // if name is repeated, add index as suffix
        const sameNameArgLen = allArgs.filter(
          existItem => existItem.name === item.name,
        ).length
        const argName = `${item.name}${sameNameArgLen || ''}`
        // save all args in allArgs array
        allArgs.push({
          ...item,
          name: argName,
        })
        return `${item.name}: $${argName}`
      })
      .join(', ')})`
  }
  const genObjectTypeStr = (type: ObjectTypeDef, _lineIndent: number) => {
    return `${type.fields.reduce((fieldsStr, field) => {
      return (
        fieldsStr +
        _genObjectFieldTypeDefStr(field, _lineIndent, indent, allArgs)
      )
    }, '')}`
  }
  switch (output.kind) {
    case 'Scalar':
    case 'Enum':
      str += '\n'
      break
    case 'Object':
      str += `${genSpace(1)}{\n${genObjectTypeStr(
        output,
        lineIndent + indent,
      )}${genSpace(lineIndent)}}\n`
      break
    case 'Union':
      str += `${genSpace(1)}{${output.types.reduce((typesStr, type) => {
        return (
          typesStr +
          `${genSpace(lineIndent + indent)}...${genSpace(1)}on${genSpace(1)}${
            type.ofName
          }${genSpace(1)}{\n${genObjectTypeStr(
            type,
            lineIndent + indent * 2,
          )}${genSpace(lineIndent + indent)}}\n`
        )
      }, '\n')}${genSpace(lineIndent)}}\n`
      break
  }
  return str
}

/**
 * generate a gql string
 * @param operation - operation which need to be converted into a gql
 * @param indent - the indent of gql
 */
export const genGQLStr = (operation: TypedOperation, indent = 2) => {
  const descriptionStr = operation.description
    ? `# ${operation.description}\n`
    : ''

  const operationType = operation.operationType
  const operationName = operation.name
  const allArgs: ArgTypeDef[] = []
  const objectFieldTypeDefStr = _genObjectFieldTypeDefStr(
    operation,
    indent,
    indent,
    allArgs,
  )
  const outerArgsStr = `(${allArgs
    .map(item => `${item.name}: ${item.type.name}`)
    .join(', ')})`
  return `${descriptionStr}${operationType}${genSpace(
    1,
  )}${operationName}${outerArgsStr}${genSpace(1)}{\n${objectFieldTypeDefStr}}`
}

/**
 * generate a group of gqls
 * @param groupName - the name of group
 * @param operations - the operations in a group
 */
export const genGQLStrInGroup = (
  groupName: string,
  operations: TypedOperation[],
) => {
  return operations.reduce((str, operation, index) => {
    const isLast = index === operations.length - 1
    return str + genGQLStr(operation) + (isLast ? '\n' : '\n\n')
  }, `# ${groupName} \n\n`)
}
