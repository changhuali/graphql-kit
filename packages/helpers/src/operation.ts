import {
  getNamedType,
  isEnumType,
  isScalarType,
  isUnionType,
  OperationTypeNode,
} from 'graphql'
import type {
  GraphQLField,
  GraphQLInputFieldMap,
  GraphQLNamedType,
  GraphQLSchema,
  GraphQLFieldMap,
} from 'graphql'
import type {
  FieldTypeDef,
  Operation,
  TypedOperation,
  TypeDef,
  OperationArgument,
  ScalarMap,
} from './interface'

/**
 * get type definition from GraphQLFieldMap or GraphQLInputFieldMap instance
 * @param fieldMap - GraphQLFieldMap or GraphQLInputFieldMap instance from which should be extracted
 */
function _getObjectTypeDefFromGraphQLFieldMap(
  fieldMap: GraphQLFieldMap<unknown, unknown> | GraphQLInputFieldMap,
) {
  const result: Record<string, FieldTypeDef> = {}
  Object.keys(fieldMap).forEach(fieldName => {
    const field = fieldMap[fieldName]
    const namedType = getNamedType(field.type)
    result[fieldName] = {
      description: field.description || '',
      type: namedType.name,
      directives: field.astNode?.directives || [],
      typeDef: _getTypeDefFromGraphQLNamedType(namedType),
    }
  })
  return result
}

/**
 * get type definition from GraphQLNamedType instance
 * @param namedType - GraphQLNamedType instance from which should be extracted
 */
function _getTypeDefFromGraphQLNamedType(namedType: GraphQLNamedType): TypeDef {
  // if the type is scalar, typeDef should be undefined
  if (isScalarType(namedType)) {
    return undefined
  }
  // if the type is enum, return an array of enum item
  if (isEnumType(namedType)) {
    return namedType.getValues().map(item => ({
      name: item.name,
      description: item.description || '',
      directives: item.astNode?.directives || [],
      value: item.value,
    }))
  }
  // if the type is union, return the object including all fields
  if (isUnionType(namedType)) {
    return namedType
      .getTypes()
      .map(item => {
        return _getObjectTypeDefFromGraphQLFieldMap(item.getFields())
      })
      .reduce((result, item) => {
        return { ...result, ...item }
      }, {})
  }
  // type should be an instance of Object here
  return _getObjectTypeDefFromGraphQLFieldMap(namedType.getFields())
}

/**
 * TODO:handle circular ref
 * get operation info from a GraphQLField instance
 * @param graphQLField - GraphQLField instance from which should be extracted
 * @param schema - GraphQLSchema instance
 */
export function getOperationFromGraphQLField(
  graphQLField: GraphQLField<unknown, unknown>,
  schema: GraphQLSchema,
): Operation {
  const typeMap = schema.getTypeMap()
  const returnTypeName = getNamedType(graphQLField.type).name
  const returnType = typeMap[returnTypeName]
  return {
    name: graphQLField?.name,
    description: graphQLField?.description || '',
    directives: graphQLField.astNode?.directives || [],
    arguments: graphQLField?.args.map(item => {
      const argTypeName = getNamedType(item.type).name
      const argType = typeMap[argTypeName]
      return {
        name: item.name,
        description: item.description || '',
        directives: item.astNode?.directives || [],
        type: argTypeName,
        typeDef: _getTypeDefFromGraphQLNamedType(argType),
      }
    }),
    return: {
      name: graphQLField.name,
      directives: returnType.astNode?.directives || [],
      description: returnType.astNode?.description?.value || '',
      type: returnTypeName,
      typeDef: _getTypeDefFromGraphQLNamedType(returnType),
    },
  }
}

/**
 * get all operations by schema
 * @param schema - GraphqlSchema instance
 */
export function getOperationsBySchema(schema: GraphQLSchema): TypedOperation[] {
  return [
    ...Object.values(schema.getQueryType()?.getFields() || {}).map(
      operationField => {
        return {
          ...getOperationFromGraphQLField(operationField, schema),
          type: OperationTypeNode.QUERY,
        }
      },
    ),
    ...Object.values(schema.getMutationType()?.getFields() || {}).map(
      operationField => {
        return {
          ...getOperationFromGraphQLField(operationField, schema),
          type: OperationTypeNode.MUTATION,
        }
      },
    ),
    ...Object.values(schema.getSubscriptionType()?.getFields() || {}).map(
      operationField => {
        return {
          ...getOperationFromGraphQLField(operationField, schema),
          type: OperationTypeNode.SUBSCRIPTION,
        }
      },
    ),
  ]
}

/**
 * grouping logic
 * @param operation - the operation need to be grouped
 */
const _groupBy: GroupByFn = (operation: TypedOperation) => {
  const [groupName, description] = operation.description.includes(':')
    ? operation.description.split(/:\s*/)
    : ['default', operation.description]
  const groupOperation = { ...operation, description }
  return { groupName, operation: groupOperation }
}

export interface GroupByFn {
  (operation: TypedOperation): { groupName: string; operation: TypedOperation }
}

/**
 * separate operations into some groups
 * @param operations - the operations need to be grouped
 * @param groupBy - the grouping logic function
 */
export function groupOperations(
  operations: TypedOperation[],
  groupBy: GroupByFn = _groupBy,
) {
  const groupMap: Record<string, TypedOperation[]> = {}
  operations.forEach(originOperation => {
    const { groupName, operation } = groupBy(originOperation)
    if (groupMap[groupName]) {
      groupMap[groupName].push(operation)
    } else {
      groupMap[groupName] = [operation]
    }
  })
  return groupMap
}

/**
 * generate a variables object
 * @param args - the arguments of operation
 * @param scalarMap - a map contains the default value of scalar type
 */
export const genVariables = (
  args: OperationArgument[],
  scalarMap: ScalarMap,
) => {
  const variables: Record<string, unknown> = {}
  args.forEach(({ name, type, typeDef }) => {
    let defaultValue
    if (!typeDef) {
      // scalar type
      const valueHandler = scalarMap[type]
      defaultValue = valueHandler
        ? typeof valueHandler === 'function'
          ? valueHandler()
          : valueHandler
        : null
    } else if (Array.isArray(typeDef)) {
      // enum type
      defaultValue = typeDef[0].value
    } else {
      // object type
      const subArgs = Object.entries(typeDef).map(
        ([fieldName, fieldTypeDef]) => {
          return {
            name: fieldName,
            ...fieldTypeDef,
          }
        },
      )
      defaultValue = genVariables(subArgs, scalarMap)
    }
    variables[name] = defaultValue
  })
  return variables
}
