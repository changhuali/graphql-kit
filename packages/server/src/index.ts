export * from './server'
export * from './interface'
