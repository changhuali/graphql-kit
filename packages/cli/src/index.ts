import { writeFile } from 'fs/promises'
import * as path from 'path'
import { existsSync } from 'fs'
import { Command } from 'commander'
import { isScalarType } from 'graphql'
import inquirer from 'inquirer'
import { watch } from 'chokidar'
import { getGraphQLSchema, startServer } from '@fruits-chain/graphql-kit-server'
import chalk from 'chalk'
import resolveDeps from 'resolve-dependencies'
import obj2str from 'stringify-object'
import { version } from '../package.json'
import { initQs } from './questions'
import type { GraphQLSchema } from 'graphql'
import type { Answers } from './questions'
import type { Server } from 'http'
import type { FSWatcher } from 'chokidar'
import type {
  GraphqlKitConfig,
  MockConfig,
} from '@fruits-chain/graphql-kit-server'
import type { FileMap } from 'resolve-dependencies/lib/file'

const program = new Command()
program.version(version)

// gen typeMapper automatically
const genTypeMapper = (typeMap: ReturnType<GraphQLSchema['getTypeMap']>) => {
  const mapper: MockConfig['typeMapper'] = {}
  Object.values(typeMap).forEach(type => {
    const typeName = type.name
    if (
      !typeName.startsWith('_') &&
      isScalarType(type) &&
      !mapper[typeName] &&
      typeName
    ) {
      mapper[typeName] = () => null
    }
  })
  return mapper
}
// create a init command
program
  .command('init')
  .description('run config initialization wizard')
  .action(() => {
    const prompt = inquirer.createPromptModule()
    prompt(initQs).then(async (answers: Answers) => {
      let typeMapper
      try {
        const graphqlSchema = await getGraphQLSchema({
          endpointUrl: answers.endpoint.url,
        })
        const typeMap = graphqlSchema.getTypeMap()
        typeMapper = genTypeMapper(typeMap)
      } catch (err) {
        console.log(chalk.red(err))
        process.exit(1)
      }
      const config = {
        ...answers,
        port: +answers.port,
        mock: {
          ...answers.mock,
          typeMapper,
        },
      }

      // gen config file
      const contentStr = `module.exports = ${obj2str(config, {
        indent: '  ',
      })}`
      await writeFile(
        path.resolve(process.cwd(), 'graphql-kit.config.js'),
        contentStr,
        'utf-8',
      )
      console.info(contentStr)
      console.info(
        `${chalk.greenBright(
          'ðŸ¤– A configuration file has been generated, now you can run ',
        )}${chalk.cyan('yarn/npm gk start')}${chalk.greenBright(
          ' to start a server',
        )}`,
      )
    })
  })

const getDepFiles = (deps: FileMap, depChain: string[]) => {
  let depFiles: string[] = []
  Object.values(deps).forEach(dep => {
    if (
      dep &&
      !dep.absPath.includes('node_modules') &&
      !depChain.includes(dep.absPath)
    ) {
      depFiles.push(dep.absPath)
      depFiles = depFiles.concat(
        getDepFiles(dep.deps, [...depChain, dep.absPath]),
      )
    }
  })
  return depFiles
}
const getConfigDepFiles = async (configPath: string) => {
  let depFiles: string[] = []
  const config = (await import(configPath)) as GraphqlKitConfig
  if (config.mock?.enable) {
    depFiles = [...(config.mock?.schemaFiles || [])]
  }
  Object.values(resolveDeps(configPath).entries).forEach(entry => {
    if (entry) {
      depFiles.push(entry.absPath)
      depFiles = depFiles.concat(getDepFiles(entry.deps, [entry.absPath]))
    }
  })
  return depFiles
}

// create a start command
program
  .command('start')
  .description('start a graphql server')
  .action(async () => {
    // validate config file
    const configPath = path.join(process.cwd(), 'graphql-kit.config.js')
    if (!existsSync(configPath)) {
      console.log(
        chalk.red(
          `${configPath} is not exist, you should use ${chalk.cyan(
            'gk init',
          )} to generate it`,
        ),
      )
      process.exit(1)
    }

    let server: Server
    let watcher: FSWatcher
    const loadServer = async () => {
      server = await startServer(configPath)
      const configDepFiles = await getConfigDepFiles(configPath)
      watcher = watch(configDepFiles, {
        ignoreInitial: true,
      }).on('all', () => {
        console.log(
          chalk.cyan('Config file has changed, server will be reload'),
        )
        server.close(async err => {
          if (!err) {
            await watcher.close()
            await loadServer()
          } else {
            console.log(chalk.red(err))
          }
        })
      })
    }

    await loadServer()
  })

program.parse(process.argv)
