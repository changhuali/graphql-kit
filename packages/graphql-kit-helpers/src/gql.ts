import { OperationTypeNode } from 'graphql'
import {
  Operation,
  OperationArgument,
  TypedOperation,
  ScalarMap,
  TypeDef,
} from './types'

/**
 * generate a space string
 * @param count the str length
 */
export const genSpace = (count: number) => {
  return ' '.repeat(count)
}

/**
 * generate an outer arguments string
 * @param args the arguments of operation
 */
const _genOuterArgumentsStr = (args: OperationArgument[]) => {
  if (args.length) {
    return `(${args.map(item => `$${item.name}: ${item.type}`).join(', ')})`
  }
  return ''
}

/**
 * generate an inner arguments string
 * @param args operation's arguments
 */
const _genInnerArgumentsStr = (args: OperationArgument[]) => {
  if (args.length) {
    return `(${args.map(item => `${item.name}: $${item.name}`).join(', ')})`
  }
  return ''
}

/**
 * generate a typeDef gql
 * @param args the type of operation
 * @param indent the indent of gql
 */
const _genTypeDefStr = (typeDef: TypeDef, indent: number) => {
  if (!typeDef) return '\n'
  let str = ''
  if (Array.isArray(typeDef)) {
    str += '\n'
  } else {
    str += ` {\n`
    const fields = Object.entries(typeDef)
    fields.forEach(([field, fieldInfo]) => {
      str += `${genSpace(indent)}${field}`
      str += _genTypeDefStr(fieldInfo.typeDef, indent + 2)
    })
    str += `${genSpace(indent - 2)}}\n`
  }
  return str
}

/**
 * generate a gql string
 * @param operationType the type of operation
 * @param operation operation which need to be converted into a gql
 * @param indent the indent of gql
 */
export const genGQLStr = (
  operationType: OperationTypeNode,
  operation: Operation,
  indent = 2,
) => {
  const returnTypeDef = operation.return.typeDef
  const descriptionStr = operation.description
    ? `# ${operation.description}\n`
    : ''
  return `${descriptionStr}${operationType}${genSpace(1)}${
    operation.name
  }${_genOuterArgumentsStr(operation.arguments)}${genSpace(1)}{\n${genSpace(
    indent,
  )}${operation.name}${_genInnerArgumentsStr(operation.arguments)}${
    returnTypeDef ? _genTypeDefStr(returnTypeDef, indent + indent) : '\n'
  }}`
}

/**
 * generate a variables object
 * @param args the arguments of operation
 * @param scalarMap a map contains the default value of scalar type
 */
export const genVariables = (
  args: OperationArgument[],
  scalarMap: ScalarMap,
) => {
  const variables: Record<string, unknown> = {}
  args.forEach(({ name, type, typeDef }) => {
    let defaultValue
    if (!typeDef) {
      // scalar type
      defaultValue = scalarMap[type]
    } else if (Array.isArray(typeDef)) {
      // enum type
      defaultValue = typeDef[0].value
    } else {
      // object type
      const subArgs = Object.entries(typeDef).map(
        ([fieldName, fieldTypeDef]) => {
          return {
            name: fieldName,
            ...fieldTypeDef,
          }
        },
      )
      defaultValue = genVariables(subArgs, scalarMap)
    }
    variables[name] = defaultValue
  })
  return variables
}

/**
 * generate a group of gqls
 * @param groupName the name of group
 * @param operations the operations in a group
 */
export const genGQLStrInGroup = (
  groupName: string,
  operations: TypedOperation[],
) => {
  return operations.reduce((str, { type, ...operation }, index) => {
    const isLast = index === operations.length - 1
    return str + genGQLStr(type, operation) + (isLast ? '\n' : '\n\n')
  }, `# ${groupName} \n\n`)
}
