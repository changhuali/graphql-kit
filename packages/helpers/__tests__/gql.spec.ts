import { readFileSync } from 'fs'
import { URL, fileURLToPath } from 'url'
import { OperationTypeNode, buildSchema } from 'graphql'
import {
  getOperationsBySchema,
  groupOperations,
  normalizeGraphqlField,
  genArgsExample,
  genOutputExample,
} from '../src/operation'
import { genSpace, genGQLStr, genGQLStrInGroup } from '../src/gql'

/********** genSpace start **********/
describe('genSpace', () => {
  it('a string consisting of a given number of spaces should be generated', () => {
    expect(genSpace(0)).toBe('')
    expect(genSpace(10)).toBe(' '.repeat(10))
  })
})
/********** genSpace end ************/

/********** genGQLStr start **********/
const schemaPath = fileURLToPath(new URL('./schema.graphql', import.meta.url))
const schemaString = readFileSync(schemaPath, {
  encoding: 'utf-8',
})
const schema = buildSchema(schemaString)
const allOperations = getOperationsBySchema(schema)
const operationsAfterBeingGrouped = groupOperations(allOperations)

describe('genGQLStr', () => {
  const fieldsMap = {
    [OperationTypeNode.QUERY]: schema.getQueryType()?.getFields() || {},
    [OperationTypeNode.MUTATION]: schema.getMutationType()?.getFields() || {},
    [OperationTypeNode.SUBSCRIPTION]:
      schema.getSubscriptionType()?.getFields() || {},
  }
  Object.entries(fieldsMap).forEach(([operationType, graphQLFields]) => {
    Object.values(graphQLFields).forEach(graphQLField => {
      it(`a gql string for ${graphQLField.name} ${operationType} should be generated`, () => {
        const operation = normalizeGraphqlField(graphQLField, {})
        const gqlString = genGQLStr({
          operationType: operationType as OperationTypeNode,
          ...operation,
        })
        expect(gqlString).toMatchSnapshot()
      })
    })
  })
})
/********** genGQLStr start ************/

/********** genGQLStrInGroup start ************/
describe('genGQLStrInGroup', () => {
  Object.entries(operationsAfterBeingGrouped).forEach(
    ([groupName, operations]) => {
      it(`gql strings for group ${groupName} should be generated`, () => {
        const gqlStrings = genGQLStrInGroup(groupName, operations)
        expect(gqlStrings).toMatchSnapshot()
      })
    },
  )
})
/********** genGQLStrInGroup end **************/

/********** genExampleValue start ************/
describe('genExampleValue', () => {
  allOperations.forEach(operation => {
    it(`a arguments example for the operation ${operation.name} ${operation.operationType} should be generated`, () => {
      expect(genArgsExample(operation.args, {})).toMatchSnapshot()
    })
    it(`a return data example for the operation ${operation.name} ${operation.operationType} should be generated`, () => {
      expect(genOutputExample(operation.output, {})).toMatchSnapshot()
    })
  })
})
/********** genExampleValue end **************/
