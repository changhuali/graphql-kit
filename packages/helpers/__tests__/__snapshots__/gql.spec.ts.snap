// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genExampleValue a arguments example for the operation role query should be generated 1`] = `
Object {
  "a": null,
  "b": null,
}
`;

exports[`genExampleValue a return data example for the operation role query should be generated 1`] = `
Object {
  "name": null,
  "sex": null,
}
`;

exports[`genGQLStr a gql string for role query should be generated 1`] = `
"query role(a: ID, b: String!, a1: ID, b1: ID) {
  role(a: $a, b: $b) {
    ... on Student {
      sex
      name
    }
    ... on Teacher {
      sex
      name
      level(a: $a1, b: $b1)
    }
  }
}"
`;

exports[`genGQLStrInGroup gql strings for group default should be generated 1`] = `
"# default 

query role(a: ID, b: String!, a1: ID, b1: ID) {
  role(a: $a, b: $b) {
    ... on Student {
      sex
      name
    }
    ... on Teacher {
      sex
      name
      level(a: $a1, b: $b1)
    }
  }
}
"
`;
